mesonbuild/modules/unstable_external_project.py

```diff
@@ -140,13 +140,30 @@ def _configure(self, state: 'ModuleState') -> None:
                 continue
             cargs = self.env.coredata.get_external_args(MachineChoice.HOST, lang)
             assert isinstance(cargs, list), 'for mypy'
+
+            cargs += compiler.get_always_args()
+            cargs += compiler.get_warn_args(T.cast(str, self.env.coredata.get_option(
+                OptionKey('warning_level', state.subproject))))
+            cargs += compiler.get_optimization_args(T.cast(
+                str, self.env.coredata.get_option(OptionKey('optimization', state.subproject))))
+            if T.cast(bool, self.env.coredata.get_option(OptionKey('werror', state.subproject))):
+                cargs += compiler.get_werror_args()
+            cargs += compiler.get_debug_args(T.cast(bool, self.env.coredata.get_option(
+                OptionKey('debug', state.subproject))))
+            cargs += compiler.get_buildtype_args(T.cast(str,
+                self.env.coredata.get_option(OptionKey('buildtype', state.subproject))))
```

Comment:
The ordering seems wrong for this. While many of these are going to be non-conflicting groups of arguments, the external args should come last in order to override anything else.

See `generate_basic_compiler_args()` in mesonbuild/backend/backends.py for the logical ordering we use there. It may also be useful to refactor some of that logic into something shareable between backends and unstable_external_project.

Reply:
Thanks for the feedback. I'll look into this tonight. Hopefully we can get it in in the upcoming release.

Reply:
So far I am thinking that the Backend class is doing too much. I think generate_basic_compiler_args() should essentially be moved to the Compiler class.

```python
def get_basic_compiler_args(self,
            env: T.Optional[Environment]=None,
            target: T.Optional[Target]=None,
            subproject: T.Optional[SubProject]=None) -> T.List[str]:
    ...
```
Requires mirroring get_option_for_target() in the Compiler class. Any thoughts on this potential approach?

Generally it feels like the Backend class is doing too much to get compiler args and should just be able to ask the Compiler class for most of it. get_source_dir_include_args() and get_build_dir_include_args() for instance.

Reply:
Maybe get_option_for_target() could be moved to the Environment class?

Reply:
@dcbaker thoughts?